x/concentrated-liquidity/math/tick.go

```diff
@@ -257,3 +274,70 @@ func CalculateSqrtPriceToTick(sqrtPrice sdk.Dec) (tickIndex int64, err error) {
 	}
 	return truncatedTick - 1, nil
 }
+
+// CalculateSqrtPriceToTick takes in a square root and returns the corresponding tick index.
+// This function does not take into consideration tick spacing.
+func CalculateSqrtPriceToTickBigDec(sqrtPrice osmomath.BigDec) (tickIndex int64, err error) {
+	// SqrtPrice may have errors, so we take the tick obtained from the price
+	// and move it in a +/- 1 tick range based on the sqrt price those ticks would imply.
+	price := sqrtPrice.Mul(sqrtPrice)
+
+	// It is acceptable to truncate price as the minimum we support is
+	// 10**-12 which is above the smallest value of sdk.Dec.
+	priceDec := price.SDKDec()
+
+	tick := CalculatePriceToTickDec(priceDec)
+	truncatedTick := tick.TruncateInt64()
+
+	// We have a candidate bucket index `t`. We discern here if:
+	// * sqrtPrice in [TickToSqrtPrice(t - 1), TickToSqrtPrice(t))
+	// * sqrtPrice in [TickToSqrtPrice(t), TickToSqrtPrice(t + 1))
+	// * sqrtPrice in [TickToSqrtPrice(t+1), TickToSqrtPrice(t + 2))
+	// * sqrtPrice not in either.
+	// We handle boundary checks, by saying that if our candidate is the min tick,
+	// set the candidate to min tick + 1.
+	// If our candidate is at or above max tick - 1, set the candidate to max tick - 2.
+	// This is because to check tick t + 1, we need to go to t + 2, so to not go over
+	// max tick during these checks, we need to shift it down by 2.
+	// We check this at max tick - 1 instead of max tick, since we expect the output to
+	// have some error that can push us over the tick boundary.
+	outOfBounds := false
+	if truncatedTick <= types.MinInitializedTick {
+		truncatedTick = types.MinInitializedTick + 1
+		outOfBounds = true
+	} else if truncatedTick >= types.MaxTick-1 {
+		truncatedTick = types.MaxTick - 2
+		outOfBounds = true
+	}
+
+	sqrtPriceTmin1, errM1 := TickToSqrtPriceBigDec(truncatedTick - 1)
+	sqrtPriceT, errT := TickToSqrtPriceBigDec(truncatedTick)
+	sqrtPriceTplus1, errP1 := TickToSqrtPriceBigDec(truncatedTick + 1)
+	sqrtPriceTplus2, errP2 := TickToSqrtPriceBigDec(truncatedTick + 2)
```

Comment:
note: these sqrt prices are truncated at precision of 18 but `sqrtPrice` has precision of 36